tokio-udp/src/recv_dgram.rs

```diff
@@ -30,6 +30,67 @@ impl<T> RecvDgram<T> {
         let inner = RecvDgramInner { socket: socket, buffer: buffer };
         RecvDgram { state: Some(inner) }
     }
+
+    /// Consume the `RecvDgram`, returning the socket.
+    ///
+    /// ```
+    /// # extern crate tokio_udp;
+    ///
+    /// use tokio_udp::UdpSocket;
+    ///
+    /// # pub fn main() {
+    ///
+    /// let socket = UdpSocket::bind(&([127, 0, 0, 1], 0).into()).unwrap();
+    /// let mut buffer = vec![0; 4096];
+    ///
+    /// let future = socket.recv_dgram(buffer);
+    ///
+    /// // ... polling `future` ... giving up (e.g. after timeout)
+    ///
+    /// let socket = future.into_inner();
+    ///
+    /// # }
+    /// ```
+    ///
+    /// # Panics
+    ///
+    /// If called after the future has completed.
+    pub fn into_inner(mut self) -> UdpSocket {
+        let state = self.state
```

Comment:
How about adding common private function to get the state?

Reply:
I am not sure it would be an improvement. It would essentially look like this:

```rust
fn into_state(mut self, msg: &str) -> RcvDgramInner {
    self.state.take().expect(msg)
}
```

and the calls would be `self.into_state("into_inner called after completion")` and `self.into_state("into_parts called after completion")` which does not really seem clearer to me. Would you really like me to make this change?